### PR TITLE
TS-3967: Set stream state to close after a RST_STEAM has been sent

### PR DESCRIPTION
**Pull request by maskit**
https://issues.apache.org/jira/browse/TS-3967

The original patches are on the JIRA ticket. I updated the patches to resolve conflicts, and made little changes to satisfy h2spec tests.

### CODE REVIEW

File: proxy/http2/RegressionHPACK.cc [hunk 1]
```diff
@@ -441,6 +441,7 @@ REGRESSION_TEST(HPACK_Decode)(RegressionTest *t, int, int *pstatus)
   box = REGRESSION_TEST_PASSED;
 
   Http2DynamicTable dynamic_table;
+  bool trailing_header;
```
Comment (masaori335) [inline proxy/http2/RegressionHPACK.cc]: Should be initialised?

File: proxy/http2/HTTP2.cc [hunk 2]
```diff
@@ -665,11 +666,19 @@ http2_decode_header_blocks(HTTPHdr *hdr, const uint8_t *buf_start, const uint8_t
       }
     }
 
+    // turn on that we have a trailer header
+    char trailer_name[] = "trailer";
+    if (name_len == (sizeof(trailer_name) - 1) && strncmp(name, trailer_name, sizeof(trailer_name) - 1) == 0) {
+      trailing_header = true;
```
Comment (masaori335) [inline proxy/http2/HTTP2.cc]: `is_trailing_header` ? It looks like `trailing_header` is not used in below code.

File: proxy/http2/HTTP2.cc [hunk 3]
```diff
@@ -665,11 +666,19 @@ http2_decode_header_blocks(HTTPHdr *hdr, const uint8_t *buf_start, const uint8_t
       }
     }
 
+    // turn on that we have a trailer header
+    char trailer_name[] = "trailer";
```
Comment (masaori335) [inline proxy/http2/HTTP2.cc]: `const` ? Should be defined in HTTP2.h ?

File: proxy/http2/HTTP2.cc (hunk 2 above)
Comment (maskit) [inline proxy/http2/HTTP2.cc]: `trailing_header` is for caller. It suggests existence of trailing header to outside of this function.
`is_trailing_header` is for callee. It suggests that this block is trailing header to skip the pseudo header checks.

Comment (masaori335) [inline proxy/http2/HTTP2.cc]: oh, I didn't notice '&' in argument. sorry:p

**Comment by maskit** [comment]
rebased to follow the changes in master

**Pull request closed without merging**